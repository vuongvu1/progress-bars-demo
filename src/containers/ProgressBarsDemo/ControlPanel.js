import React, { useState, useEffect, useCallback } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

import ProgressBar from 'components/ProgressBar';

// #region Styles
const Wrapper = styled.div`
  width: 400px;
  background-color: ${(props) => props.theme.lightBlue};
  padding: 20px;
  display: flex;
  justify-content: center;
  align-items: center;
  flex-wrap: wrap;
`;

const Title = styled.div`
  font-size: ${(props) => props.theme.fontSizeBig};
  margin-bottom: 20px;
`;

const BarsWrapper = styled.div`
  width: 100%;

  * {
    margin-bottom: 10px;
  }
`;

const ControlPanelWrapper = styled.div`
  width: 100%;
  text-align: center;

  * {
    margin-right: 5px;
  }
`;
// #endregion Styles


const ControlPanel = ({ barList, buttonList, limitValue }) => {
  const [activeBar, setActiveBar] = useState('');
  const [bars, setBars] = useState(barList);
  const [buttons, setButtons] = useState(buttonList);

  useEffect(() => { setBars(barList); }, [barList]);
  useEffect(() => { setButtons(buttonList); }, [buttonList]);

  const changeValue = useCallback((value) => {
    const newBars = bars.map((bar) => {
      if (activeBar === bar.id) {
        const newValue = bar.data + value;
        return {
          ...bar,
          data: newValue < 0 ? 0 : newValue,
        };
      }
      return bar;
    });
    setBars(newBars);
  }, [activeBar, bars]);

  return (
    <Wrapper>
      <Title>
        {activeBar
          ? `Selected (${activeBar})`
          : 'Select a bar to update'}
      </Title>
      <BarsWrapper>
        {bars.map(({ id, data }) => (
          <ProgressBar
            key={id}
            id={id}
            value={data}
            limit={limitValue}
            isActive={activeBar === id}
            setActive={setActiveBar}
          />
        ))}
      </BarsWrapper>

      <ControlPanelWrapper>
        {buttons.map(({ id, data }) => (
          <input
            key={id}
            type='button'
            value={data > 0 ? `+${data}` : data}
            onClick={() => changeValue(data)}
          />
        ))}
      </ControlPanelWrapper>
    </Wrapper>
  );
};

ControlPanel.defaultProps = {
  barList: [],
  buttonList: [],
  limitValue: 0,
};

ControlPanel.propTypes = {
  barList: PropTypes.arrayOf(PropTypes.shape({
    id: PropTypes.string,
    data: PropTypes.number,
  })),
  buttonList: PropTypes.arrayOf(PropTypes.shape({
    id: PropTypes.string,
    data: PropTypes.number,
  })),
  limitValue: PropTypes.number,
};

export default ControlPanel;
