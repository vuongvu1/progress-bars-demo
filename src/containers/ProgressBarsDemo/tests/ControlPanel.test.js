import React from 'react';
import { render } from '@testing-library/react';

import ControlPanel from '../ControlPanel';

const renderComponent = (props = {}) => render(<ControlPanel {...props} />);

describe('<ControlPanel />', () => {
  it('should render <ControlPanel />', () => {
    const { container } = renderComponent({
      barList: [{ id: 'id_0', data: 45 }, { id: 'id_1', data: 20 }, { id: 'id_2', data: 80 }],
      buttonList: [{ id: 'id_0', data: 54 }, { id: 'id_1', data: 13 }, { id: 'id_2', data: -26 }],
      limitValue: 170,
    });
    expect(container).toMatchSnapshot();
  });
});
