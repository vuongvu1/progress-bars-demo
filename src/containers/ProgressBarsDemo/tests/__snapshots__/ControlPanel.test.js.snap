// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ControlPanel /> should render <ControlPanel /> 1`] = `
<div>
  <div
    class="sc-bxivhb ksgGse"
  >
    <div
      class="sc-ifAKCX jdJDTU"
    >
      Click to select a bar
    </div>
    <div
      class="sc-EHOje gNssoU"
    >
      <div
        class="sc-bdVaJa edJoaQ"
      >
        <div>
          26
          %
        </div>
        <div
          class="sc-bwzfXH gQkkZY"
          value="26"
        />
      </div>
      <div
        class="sc-bdVaJa edJoaQ"
      >
        <div>
          12
          %
        </div>
        <div
          class="sc-bwzfXH ieWoee"
          value="12"
        />
      </div>
      <div
        class="sc-bdVaJa edJoaQ"
      >
        <div>
          47
          %
        </div>
        <div
          class="sc-bwzfXH fNhdMA"
          value="47"
        />
      </div>
    </div>
    <div
      class="sc-bZQynM fgWUIq"
    >
      <button
        class="sc-htpNat bXUvhW"
      >
        +54
      </button>
      <button
        class="sc-htpNat bXUvhW"
      >
        +13
      </button>
      <button
        class="sc-htpNat bXUvhW"
      >
        -26
      </button>
    </div>
  </div>
</div>
`;
